composition-js/src/merging/merge.ts

```diff
@@ -1810,79 +1857,64 @@ class Merger {
     //   definition is the intersection of all definitions (meaning that if there divergence in
     //   locations, we only expose locations that are common everywhere).
     this.mergeDescription(sources, dest);
-    if (sources.some((s) => s && this.isMergedDirective(s))) {
+
+    const definedSources = sources.filter((src): src is DirectiveDefinition => src !== undefined);
+    if (definedSources.some(s => s.locations.some(loc => executableDirectiveLocations.includes(loc)))) {
+      if (this.mergedDirectives().includes(dest.name)) {
+        const { locations } = getLocationsFromDirectiveDefs(definedSources);
+        dest.addLocations(...locations);
+      }
       this.mergeExecutableDirectiveDefinition(sources, dest);
+    } else {
+      if (this.mergedDirectives().includes(dest.name)) {
```

Comment:
Earlier `isMergedDirective` calls should have ensured that this test is not necessary? I believe the same test in the previous branch is also unnecessary, though for the slightly different reason that `getLocationsFromDirectiveDefs` should be safe to call for executable-only directives.

 In general, I'd find it "reassuring" if `this.mergedDirectives()` was checked only in `isMergedDirective`.

Reply:
This check does have to be here because the federation directives already have their arguments added, so removing it completely would blow up. I did change it to call `isMergeDirective` rather than repeating the same logic, which I think is a good change.